kafka-source/src/main/java/org/springframework/cloud/stream/module/kafka/source/KafkaSource.java

```diff
@@ -0,0 +1,153 @@
+import org.springframework.integration.dsl.IntegrationFlows;
+import org.springframework.integration.dsl.kafka.Kafka;
+import org.springframework.integration.kafka.core.*;
+import org.springframework.integration.kafka.listener.KafkaMessageListenerContainer;
+import org.springframework.integration.kafka.listener.KafkaNativeOffsetManager;
+import org.springframework.integration.kafka.support.ZookeeperConnect;
+import org.springframework.retry.support.RetryTemplate;
+import org.springframework.util.Assert;
+import org.springframework.util.ObjectUtils;
+import org.springframework.util.StringUtils;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Source module that receives data from Kafka.
+ *
+ * @author Soby Chacko
+ */
+@EnableBinding(Source.class)
+@EnableConfigurationProperties(KafkaConfigurationProperties.class)
+public class KafkaSource {
+
+	@Autowired
+	@Bindings(KafkaSource.class)
+	private Source source;
+
+	@Autowired
+	private KafkaConfigurationProperties properties;
+
+	private ConnectionFactory kafkaConnectionFactory;
+
+	@Bean
+	public KafkaMessageListenerContainer container() {
+
+		Assert.isTrue(!ObjectUtils.isEmpty(properties.getTopics()) || !ObjectUtils.isEmpty(properties.getPartitions()),
+				"Either a list of topics OR a topic=<comma separated partitions> must be provided");
+
+		ZookeeperConnect zookeeperConnect = new ZookeeperConnect();
+		zookeeperConnect.setZkConnect(properties.getZkConnect());
+		zookeeperConnect.setZkConnectionTimeout(properties.getZkConnectionTimeout());
+		zookeeperConnect.setZkSessionTimeout(properties.getZkSessionTimeout());
+		zookeeperConnect.setZkSyncTime(properties.getZkSyncTime());
+
+		Configuration zookeeperConfiguration = new ZookeeperConfiguration(zookeeperConnect);
+		this.kafkaConnectionFactory = new DefaultConnectionFactory(zookeeperConfiguration);
+
+		Partition[] partitions = getPartitions();
+
+		KafkaNativeOffsetManager kafkaNativeOffsetManager = new KafkaNativeOffsetManager(kafkaConnectionFactory, zookeeperConnect,
+				getInitialOffsets());
+		kafkaNativeOffsetManager.setRetryTemplate(new RetryTemplate());
+
+		KafkaMessageListenerContainer container = ObjectUtils.isEmpty(partitions) ?
+				new KafkaMessageListenerContainer(kafkaConnectionFactory, properties.getTopics()) :
+				new KafkaMessageListenerContainer(kafkaConnectionFactory, partitions);
+
+		container.setOffsetManager(kafkaNativeOffsetManager);
+
+		return container;
+	}
+
+	@Bean
+	public IntegrationFlow fromKafka() throws IllegalAccessException, InstantiationException, ClassNotFoundException {
+
+		return IntegrationFlows.from(Kafka.messageDriverChannelAdapter(container())
+				.keyDecoder(getKeyDecoder())
+				.payloadDecoder(getPayloadDecoder()))
+				.channel(source.output())
+				.get();
+	}
+
+	private Decoder<?> getKeyDecoder() throws ClassNotFoundException, IllegalAccessException, InstantiationException {
```

Comment:
@ilayaperumalg the decoders may not be the same. It was a typo that we used the KeyDecoder for both key and payload. Thanks a lot for catching this. We have separate properties for key and payload decoders in the KafkaConfigurationProperties class. I will change the code to reflect that. 
For example, Key could be a LongDecoder and the payload could be based on String.